File: chat/uikit/lib/src/video_recorder/video_recorder.dart
import 'package:tencent_chat_uikit/tencent_chat_uikit.dart';
import 'package:flutter/material.dart';

import 'video_recorder_platform.dart';

enum RecordMode {
  mixed,
  photoOnly,
  videoOnly,
}

enum VideoRecordQuality {
  low,
  medium,
  high,
}

enum RecordMediaType {
  photo,
  video,
}

class VideoRecorderResult {
  final RecordMediaType mediaType;
  final String filePath;
  final int? durationMs;
  final String? thumbnailPath;

  VideoRecorderResult({
    required this.mediaType,
    required this.filePath,
    this.durationMs,
    this.thumbnailPath,
  });

  @override
  String toString() {
    if (mediaType == RecordMediaType.photo) {
      return 'VideoRecorderResult(mediaType: photo, filePath: $filePath)';
    } else {
      return 'VideoRecorderResult(mediaType: video, filePath: $filePath, durationMs: $durationMs, thumbnailPath: $thumbnailPath)';
    }
  }
}

class VideoRecorderConfig {
  final RecordMode? recordMode;
  final VideoRecordQuality? videoQuality;
  final int? minDurationMs;
  final int? maxDurationMs;
  final bool? isDefaultFrontCamera;
  final bool? isSupportTorch;

  const VideoRecorderConfig({
    this.recordMode,
    this.videoQuality,
    this.minDurationMs,
    this.maxDurationMs,
    this.isDefaultFrontCamera,
    this.isSupportTorch,
  });
}

class VideoRecorder {
  static final VideoRecorder instance = VideoRecorder._internal();

  VideoRecorder._internal();

  static Future<VideoRecorderResult> startRecord({
    required BuildContext context,
    VideoRecorderConfig? config,
  }) async {
    if (!await Permission.checkAndRequest(context, [PermissionType.camera])) {
      throw Exception('Camera permission denied');
    }

    if (!context.mounted) {
      throw Exception('Context is no longer mounted');
    }

    final primaryColor = Theme.of(context).colorScheme.primary.toARGB32();
    return VideoRecorderPlatform.startRecordNative(
      config: config ?? const VideoRecorderConfig(),
      locale: Localizations.localeOf(context),
      primaryColor:
          '#${(primaryColor & 0xFFFFFF).toRadixString(16).padLeft(6, '0')}',
    );
  }
}

File: chat/uikit/android/src/main/kotlin/com/tencent/cloud/tuikit/flutter/tuichatkit/videorecorder/VideoRecorder.kt
package com.tencent.cloud.tuikit.flutter.tuichatkit.videorecorder
import com.tencent.cloud.tuikit.flutter.tuichatkit.videorecorder.impl.VideoRecorderViewImpl

object VideoRecorder {
    private val instance = VideoRecorderViewImpl()

    /**
     * Launches the camera capture interface
     * @param config Capture configuration settings
     * @param listener Callback interface for receiving capture results, including:
     *                 - Photo capture success
     *                 - Video recording success
     *
     * 启动相机拍摄界面 @param config 拍摄配置设置 @param listener 接收拍摄结果的回调接口，包括：- 照片拍摄成功 - 视频录制成功
     */
    fun startRecord(config: VideoRecorderConfig = VideoRecorderConfig(), listener: VideoRecordListener) {
        instance.takeVideo(config, listener)
    }
}

data class VideoRecorderConfig(
    var recordMode: RecordMode? = null,
    var videoQuality: VideoQuality? = null,
    var minDurationMs: Int? = null,
    var maxDurationMs: Int? = null,
    var isDefaultFrontCamera: Boolean? = null,
    var isSupportEdit: Boolean? = null,
    var isSupportBeauty: Boolean? = null,
    var isSupportRecordScrollFilter: Boolean? = null,
    var isSupportTorch: Boolean? = null,
    var isSupportAspect:Boolean? = null
)

enum class RecordMode(val value: Int) {
    MIXED(0),
    PHOTO_ONLY(1),
    VIDEO_ONLY(2);

    companion object {
        fun fromInteger(value: Int): RecordMode {
            return RecordMode.entries.find { it.value == value } ?: MIXED
        }
    }
}

interface VideoRecordListener {
    fun onPhotoCaptured(filePath: String?) {}

    fun onVideoCaptured(filePath: String?, durationMs: Int, thumbnailPath: String?) {}
}

enum class VideoQuality(val value: Int) {
    LOW(1),
    MEDIUM(2),
    HIGH(3);

    companion object {
        fun fromInteger(value: Int): VideoQuality {
            return VideoQuality.entries.find { it.value == value } ?: LOW
        }
    }
}